File: app/src/main/java/de/tum/in/tumcampusapp/activities/TransportationDetailsActivity.java
package de.tum.in.tumcampusapp.activities;

import android.app.AlertDialog;
import android.content.Intent;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.LinearLayout;

import com.google.gson.Gson;

import java.util.List;

import de.tum.in.tumcampusapp.R;
import de.tum.in.tumcampusapp.activities.generic.ActivityForLoadingInBackground;
import de.tum.in.tumcampusapp.auxiliary.DepartureView;
import de.tum.in.tumcampusapp.auxiliary.NetUtils;
import de.tum.in.tumcampusapp.managers.RecentsManager;
import de.tum.in.tumcampusapp.managers.TransportManager;
import de.tum.in.tumcampusapp.models.efa.Departure;
import de.tum.in.tumcampusapp.models.efa.StationResult;

/**
 * Activity to show transport departures for a specified station
 * <p>
 * NEEDS: EXTRA_STATION set in incoming bundle (station name)
 */
public class TransportationDetailsActivity extends ActivityForLoadingInBackground<String, List<Departure>> {
    public static final String EXTRA_STATION = "station";
    public static final String EXTRA_STATION_ID = "stationID";

    private LinearLayout mViewResults;
    private RecentsManager recentsManager;
    private TransportManager transportManager;
    private Gson gson;

    public TransportationDetailsActivity() {
        super(R.layout.activity_transportation_detail);
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        // get all stations from db
        recentsManager = new RecentsManager(this, RecentsManager.STATIONS);
        transportManager = new TransportManager(this);
        gson = new Gson();
        mViewResults = this.findViewById(R.id.activity_transport_result);

        Intent intent = getIntent();
        if (intent == null) {
            finish();
            return;
        }
        String location = intent.getStringExtra(EXTRA_STATION);
        setTitle(location);
        String locationID = intent.getStringExtra(EXTRA_STATION_ID);

        startLoading(location, locationID);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_transport, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == R.id.action_transport_usage) {
            new AlertDialog.Builder(this)
                    .setTitle(R.string.transport_action_usage)
                    .setMessage(R.string.transport_help_text)
                    .setPositiveButton(android.R.string.ok, null)
                    .create()
                    .show();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    /**
     * Load departure times
     *
     * @param arg Station name
     * @return List of departures
     */
    @Override
    protected List<Departure> onLoadInBackground(String... arg) {
        final String location = arg[0];
        final String locationID = arg[1];
        final StationResult stationResult = new StationResult(location, locationID, Integer.MAX_VALUE); // Quality is always 100% hit
        final String jsonStationResult = gson.toJson(stationResult);

        // save clicked station into db
        recentsManager.replaceIntoDb(jsonStationResult);

        // Check for internet connectivity
        if (!NetUtils.isConnected(this)) {
            showNoInternetLayout();
            return null;
        }

        // get departures from website
        List<Departure> departureCursor = TransportManager.getDeparturesFromExternal(this, locationID);
        if (departureCursor.isEmpty()) {
            showError(R.string.no_departures_found);
        }

        return departureCursor;
    }

    /**
     * Adds a new {@link DepartureView} for each departure entry
     *
     * @param result List of departures
     */
    @Override
    protected void onLoadFinished(List<Departure> result) {
        showLoadingEnded();
        if (result == null) {
            return;
        }
        mViewResults.removeAllViews();
        for (Departure d : result) {
            DepartureView view = new DepartureView(this, true);

            view.setOnClickListener(v -> {
                DepartureView departureView = (DepartureView) v;
                String symbol = departureView.getSymbol();
                boolean highlight;
                if (transportManager.isFavorite(symbol)) {
                    transportManager.deleteFavorite(symbol);
                    highlight = false;
                } else {
                    transportManager.addFavorite(symbol);
                    highlight = true;
                }

                // Update the other views with the same symbol
                for (int i = 0; i < mViewResults.getChildCount(); i++) {
                    DepartureView child = (DepartureView) mViewResults.getChildAt(i);
                    if (child.getSymbol()
                             .equals(symbol)) {
                        child.setSymbol(symbol, highlight);
                    }
                }
            });

            if (transportManager.isFavorite(d.getSymbol())) {
                view.setSymbol(d.getSymbol(), true);
            } else {
                view.setSymbol(d.getSymbol(), false);
            }

            view.setLine(d.getDirection());
            view.setTime(d.getDepartureTime());
            mViewResults.addView(view);
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        for (int i = 0; i < mViewResults.getChildCount(); i++) {
            View view = mViewResults.getChildAt(i);
            if (!(view instanceof DepartureView)) {
                continue;
            }
            ((DepartureView) view).removeAllCallbacksAndMessages();
        }
    }
}

File: app/src/main/java/de/tum/in/tumcampusapp/component/ui/transportation/TransportationDetailsActivity.kt
package de.tum.`in`.tumcampusapp.component.ui.transportation

import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.widget.LinearLayout
import androidx.appcompat.app.AlertDialog
import com.google.gson.Gson
import de.tum.`in`.tumcampusapp.R
import de.tum.`in`.tumcampusapp.component.other.general.RecentsDao
import de.tum.`in`.tumcampusapp.component.other.general.model.Recent
import de.tum.`in`.tumcampusapp.component.other.generic.activity.ProgressActivity
import de.tum.`in`.tumcampusapp.component.ui.transportation.model.efa.Departure
import de.tum.`in`.tumcampusapp.component.ui.transportation.model.efa.StationResult
import de.tum.`in`.tumcampusapp.database.TcaDb
import de.tum.`in`.tumcampusapp.utils.Utils
import de.tum.`in`.tumcampusapp.utils.plusAssign
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import javax.inject.Inject

/**
 * Activity to show transport departures for a specified station
 *
 *
 * NEEDS: EXTRA_STATION set in incoming bundle (station name)
 */
class TransportationDetailsActivity : ProgressActivity<Unit>(R.layout.activity_transportation_detail) {

    private val resultsView: LinearLayout by lazy {
        findViewById<LinearLayout>(R.id.activity_transport_result)
    }

    private val gson: Gson by lazy { Gson() }

    @Inject
    lateinit var database: TcaDb

    @Inject
    lateinit var transportController: TransportController

    private val disposable = CompositeDisposable()

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        injector.inject(this)

        val intent = intent
        if (intent == null) {
            finish()
            return
        }
        val location = intent.getStringExtra(EXTRA_STATION)
        title = location
        val locationID = intent.getStringExtra(EXTRA_STATION_ID)

        showLoadingStart()
        loadDetails(location, locationID)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_transport, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.action_transport_usage) {
            val dialog = AlertDialog.Builder(this)
                    .setTitle(R.string.transport_action_usage)
                    .setMessage(R.string.transport_help_text)
                    .setPositiveButton(android.R.string.ok, null)
                    .create()

            dialog.window?.setBackgroundDrawableResource(R.drawable.rounded_corners_background)
            dialog.show()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    private fun loadDetails(location: String, locationID: String) {
        // Quality is always 100% hit
        val stationResult = StationResult(location, locationID, Integer.MAX_VALUE)
        val jsonStationResult = gson.toJson(stationResult)

        // save clicked station into db
        database.recentsDao().insert(Recent(jsonStationResult, RecentsDao.STATIONS))

        disposable += transportController.fetchDeparturesAtStation(locationID)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::displayResults) {
                    // Something went wrong
                    Utils.log(it)
                    showError(R.string.no_departures_found)
                }
    }

    /**
     * Adds a new [DepartureView] for each departure entry
     *
     * @param results List of departures
     */
    private fun displayResults(results: List<Departure>?) {
        showLoadingEnded()
        if (results == null || results.isEmpty()) {
            showError(R.string.no_departures_found)
            return
        }
        resultsView.removeAllViews()
        for ((_, direction, lineSymbol, _, departureTime) in results) {
            val view = DepartureView(this, true)

            view.setOnClickListener { v ->
                val departureView = v as DepartureView
                val symbol = departureView.symbol
                val highlight = if (transportController.isFavorite(symbol)) {
                    transportController.deleteFavorite(symbol)
                    false
                } else {
                    transportController.addFavorite(symbol)
                    true
                }

                // Update the other views with the same symbol
                for (i in 0 until resultsView.childCount) {
                    val child = resultsView.getChildAt(i) as DepartureView
                    if (child.symbol == symbol) {
                        child.setSymbol(symbol, highlight)
                    }
                }
            }

            if (transportController.isFavorite(lineSymbol)) {
                view.setSymbol(lineSymbol, true)
            } else {
                view.setSymbol(lineSymbol, false)
            }

            view.setLine(direction)
            view.setTime(departureTime)
            resultsView.addView(view)
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        for (i in 0 until resultsView.childCount) {
            val view = resultsView.getChildAt(i) as? DepartureView ?: continue
            view.removeAllCallbacksAndMessages()
        }
    }

    override fun onStop() {
        super.onStop()
        disposable.clear()
    }

    companion object {
        const val EXTRA_STATION = "station"
        const val EXTRA_STATION_ID = "stationID"
    }
}